components/brave_rewards/browser/rewards_protocol_handler.cc

```diff
@@ -133,17 +62,71 @@ namespace brave_rewards {
 
 void HandleRewardsProtocol(const GURL& url,
                            content::WebContents::OnceGetter web_contents_getter,
-                           ui::PageTransition page_transition,
-                           bool has_user_gesture) {
-  DCHECK(url.SchemeIs(kRewardsScheme));
+                           ui::PageTransition page_transition) {
+  CHECK(url.SchemeIs(kRewardsScheme));
   content::GetUIThreadTaskRunner({})->PostTask(
       FROM_HERE,
       base::BindOnce(&LoadRewardsURL, url, std::move(web_contents_getter),
-                     page_transition, has_user_gesture));
+                     page_transition));
 }
 
 bool IsRewardsProtocol(const GURL& url) {
   return url.SchemeIs(kRewardsScheme);
 }
 
+absl::optional<GURL> ValidateWalletProviderRedirect(const GURL& referrer_url,
+                                                    const GURL& redirect_url) {
+  if (!referrer_url.is_valid() || !redirect_url.is_valid()) {
+    return absl::nullopt;
+  }
+
+  static const auto kAllowedReferrerUrls{[] {
+    std::map<std::string, std::vector<GURL>> allowed_urls{
+        {"bitflyer",
+         {GURL(BUILDFLAG(BITFLYER_PRODUCTION_URL)),
+          GURL(BUILDFLAG(BITFLYER_SANDBOX_URL))}},
+        {"gemini",
+         {GURL(BUILDFLAG(GEMINI_PRODUCTION_OAUTH_URL)),
+          GURL(BUILDFLAG(GEMINI_SANDBOX_OAUTH_URL))}},
+        {"uphold",
+         {GURL(BUILDFLAG(UPHOLD_PRODUCTION_OAUTH_URL)),
+          GURL(BUILDFLAG(UPHOLD_SANDBOX_OAUTH_URL))}},
+        {"zebpay",
+         {GURL(BUILDFLAG(ZEBPAY_PRODUCTION_OAUTH_URL)),
+          GURL(BUILDFLAG(ZEBPAY_SANDBOX_OAUTH_URL))}}};
+
+    for (const auto& [wallet_provider, urls] : allowed_urls) {
+      DCHECK(base::ranges::none_of(
+          urls,
+          [](const GURL& url) { return !url.is_valid() || !url.has_host(); }))
+          << wallet_provider << " has malformed referrer URL(s)!";
+    }
+
+    return allowed_urls;
+  }()};
+
+  std::string wallet_provider;
+  const auto redirect_path_segments =
+      base::SplitString(redirect_url.path_piece(), "/", base::TRIM_WHITESPACE,
+                        base::SPLIT_WANT_NONEMPTY);
+  if (!redirect_path_segments.empty()) {
+    wallet_provider = redirect_path_segments[0];
+  }
+
+  if (base::ranges::none_of(
+          kAllowedReferrerUrls.contains(wallet_provider)
+              ? kAllowedReferrerUrls.at(wallet_provider)
+              : std::vector<GURL>{},
+          [&](base::StringPiece host_piece) {
+            return referrer_url.DomainIs(host_piece);
+          },
+          &GURL::host_piece)) {
+    LOG(ERROR) << referrer_url.spec() << " was trying to redirect to "
```

Comment:
That's a good point. Although the things that would be logged there should be worthless without the client secret, I suppose. But you're right, let's get rid of those. How about logging only the hosts then?

Reply:
sgtm

Reply:
@diracdeltas I ended up logging
```cpp
LOG(ERROR) << referrer_url.host_piece() << " was trying to redirect to "
           << redirect_url.scheme_piece() << ":"
           << redirect_url.path_piece() << ", but it's not allowed.";
```
since `GURL` has some trouble parsing our custom (`rewards://`) URL (because there's no host).

Reply:
E.g.
```
[14770:259:0825/120558.383094:ERROR:rewards_protocol_handler.cc(83)] example.com was trying to redirect to rewards://zebpay/authorization, but it's not allowed.
```
on a `http://example.com/* → rewards://zebpay/authorization?error=invalid_request&error_description=User%20does%20not%20meet%20minimum%20requirements` redirect.